xdelta3-dir-patcher

```diff
@@ -21,6 +49,355 @@ class AttributeDict(dict):
+
+    def close(self):
+        self._close_archive()
+
+    @staticmethod
+    def can_open(archive):
+        return path.isfile(archive) and tarfile.is_tarfile(archive)
+
+    def list_files(self):
+        if self.items:
+            return self.items
+
+        items = []
+
+        print('Tar: Gathering filelist')
+        members = self.archive_object.getmembers()
+
+        # Lookup all member objects
+        for member in members:
+            # For directories, add a separator to the end
+            suffix = path.sep if member.isdir() else ''
+
+            items.append('%s%s' % (member.name, suffix))
+
+        # Append any directories that weren't part of the archive
+        for item in items:
+            dir_name = path.dirname(item) + path.sep
+            if len(dir_name) > 1 and not dir_name in items:
+                print('Post-adding %s to members' % dir_name)
+                items.append(dir_name)
+
+        print('Tar: Gathering completed')
+
+        self.items = items
+
+        return self.items
+
+    def expand(self, filename, extraction_path):
+        member = None
+        try:
+            member = self.archive_object.getmember(filename)
+        except KeyError as key_error:
+            if filename.endswith(path.sep):
+                try:
+                    folder_member_name = filename.rstrip(path.sep)
+                    member = self.archive_object.getmember(folder_member_name)
+                    if not member.isdir():
+                        member = None
+                except KeyError as key_error:
+                    print('No folder. Assuming its added as a placeholder',
+                          filename)
+
+                folder_path = path.join(extraction_path,
+                                       filename)
+                makedirs(folder_path, exist_ok = True)
+
+                # TODO: Test me
+                # If we have a folder in the archive, copy its
+                # permissions/stat/mode to the target (but don't expand it)
+                if member:
+                    chmod(folder_path, member.mode)
+                    utime(folder_path, (member.mtime, member.mtime))
+
+                    try:
+                        chown(folder_path, member.uid, member.gid)
+                    except PermissionError as pe:
+                        pass
+
+                return
+
+        makedirs(extraction_path, exist_ok = True)
+
+        # XXX: Not thread safe http://bugs.python.org/issue23649
+        try:
+            super()._acquire_lock()
+
+            self.archive_object.extract(member, extraction_path)
+
+            super()._release_lock()
+        except FileExistsError as fee:
+            # XXX tarfile also might have thread issues even with the lock
+            #     in place so if we get errors as reported on the bugtracker
+            #     (http://bugs.python.org/issue23649) we try to re-run the task.
+            # TODO: Test this (somehow)
```

Comment:
Couple more thoughts on the locking. I believe that each thread should have a unique member that it's trying to extract and not be attempting to be reentrant. In that case, I think that the need for a RLock is actually hiding a bug.

I also think you probably do need a try/except here so that you can release the lock on extract exceptions, but I don't think it should wrap the lock acquisition since then you'd try to release a lock you don't have. I think the code might be best like this:
```
super()._acquire_lock()
try:
    self.archive_object.extract(member, extraction_path)
finally:
    super()._release_lock()
```
Then the lock is always released and exceptions on extractions are simply passed up.